File: app/lib/streak.ts
const KEY = "tessera:streak";

export type Streak = { current: number; max: number; lastWon: number };

export function readStreak(): Streak {
  if (typeof window === "undefined") return { current: 0, max: 0, lastWon: 0 };
  try {
    const raw = window.localStorage.getItem(KEY);
    if (!raw) return { current: 0, max: 0, lastWon: 0 };
    const s = JSON.parse(raw) as Streak;
    return { current: s.current ?? 0, max: s.max ?? 0, lastWon: s.lastWon ?? 0 };
  } catch {
    return { current: 0, max: 0, lastWon: 0 };
  }
}

export function recordWin(num: number): Streak {
  const prev = readStreak();
  let current: number;
  if (prev.lastWon === num) current = prev.current; // already won today; no-op
  else if (prev.lastWon === num - 1) current = prev.current + 1;
  else current = 1;
  const next: Streak = {
    current,
    max: Math.max(prev.max, current),
    lastWon: num,
  };
  try {
    window.localStorage.setItem(KEY, JSON.stringify(next));
  } catch {}
  return next;
}

// Streak is "live" only if lastWon is today or yesterday.
export function visibleCurrent(s: Streak, today: number): number {
  if (s.lastWon === today || s.lastWon === today - 1) return s.current;
  return 0;
}

File: mobile/lib/src/streak.dart
// Port of app/lib/streak-compute.ts. A streak is a run of consecutive
// puzzle numbers; callers pass won puzzle numbers only (revealed results
// are not wins). Fixture-locked against test/fixtures/parity.json.

class Streak {
  const Streak({
    required this.current,
    required this.max,
    required this.lastWon,
  });

  final int current;
  final int max;
  final int lastWon;

  @override
  bool operator ==(Object other) =>
      other is Streak &&
      other.current == current &&
      other.max == max &&
      other.lastWon == lastWon;

  @override
  int get hashCode => Object.hash(current, max, lastWon);

  @override
  String toString() =>
      'Streak(current: $current, max: $max, lastWon: $lastWon)';
}

/// Derives a streak from won puzzle numbers.
Streak computeStreak(List<int> nums, [int importedMax = 0]) {
  final uniq = nums.toSet().toList()..sort();
  var max = 0;
  var run = 0;
  int? prev;
  for (final n in uniq) {
    run = (prev != null && n == prev + 1) ? run + 1 : 1;
    if (run > max) max = run;
    prev = n;
  }
  if (uniq.isEmpty) {
    return Streak(current: 0, max: importedMax, lastWon: 0);
  }
  return Streak(
    current: run,
    max: max > importedMax ? max : importedMax,
    lastWon: uniq.last,
  );
}

/// A streak is "live" only if its last win was today or yesterday;
/// otherwise the visible current count is 0. Port of `visibleCurrent`
/// in app/lib/streak.ts. [today] is today's puzzle number.
int visibleCurrent(Streak s, int today) {
  if (s.lastWon == today || s.lastWon == today - 1) return s.current;
  return 0;
}

/// Merge a local and a server streak: the fresher lastWon decides `current`,
/// maxima combine.
Streak mergeStreaks(Streak a, Streak b) {
  final fresher = a.lastWon >= b.lastWon ? a : b;
  return Streak(
    current: fresher.current,
    max: a.max > b.max ? a.max : b.max,
    lastWon: fresher.lastWon,
  );
}
